### PR TITLE
Improvements regarding thesis

### LINKED ISSUE #4: Instrumentation constructs wrong try body
For some files, e.g. [here](https://github.com/python-telegram-bot/python-telegram-bot/blob/master/telegram/ext/filters.py) the instrumenter selects the wrong body for the try / catch dynapyt block.
A potential fix can be found in this https://github.com/FlxB2/DynaPyt/commit/1dd3bd9535576a520f3d70ca4112a87c684044e6  commit CodeInstrumenter.py line 104.

### PR DESCRIPTION
**Pull request by FlxB2**
Adds several improvements which I'm using for a while regarding my thesis.

List of changes:
- Allows loading analyses from external packages via `--module`
- Adds option to instrument files in another directory instead of overwriting files
- Fixes #4 
- Adds `function_iid` to `_return_` so that `function_exit` in `_return_` can be called with function iid
- `pre_call` hook now contains the callable as an argument to facilitate analyses which e.g. need the function name without having to extract the AST
- Inserts absolute path to instrumented files instead of using `__file__` which returns the location of `runtime.py`
- Support for python 3.6
- Extended .gitignore for commonly used IDEs
- More detailed requirements.txt
- Removal of testfiles

### CODE REVIEW

File: src/dynapyt/runtime.py [hunk 1]
```diff
@@ -9,11 +9,8 @@ def set_analysis(new_analysis):
     analysis = new_analysis
 
 def call_if_exists(f, *args):
-    try:
-        func = getattr(analysis, f)
-        return func(*args)
-    except AttributeError:
-        return
+    func = getattr(analysis, f, lambda *args: None)
```
Comment (AryazE) [inline src/dynapyt/runtime.py:12]: I didn't find a good description of why you need to have the lambda function as default. Maybe you can add some comments explaining this in the code for future references.

**Review by AryazE — COMMENTED** [review]
LGTM
Thanks!

File: src/dynapyt/instrument/CodeInstrumenter.py [hunk 2]
```diff
@@ -92,17 +93,18 @@ def visit_AsName(self, node):
...
         get_ast = cst.SimpleStatementLine(body=[cst.Assign(targets=[cst.AssignTarget(cst.Name('_dynapyt_ast_'))], value=parse_to_ast)])
         dynapyt_imports = [cst.Newline(value='\n')]
         dynapyt_imports.append(self.__create_import(["_catch_"]))
         import_names = list(self.to_import)
         for i in range(len(updated_node.body)):
             if m.matches(updated_node.body[i], m.SimpleStatementLine()) and \
-            (m.matches(updated_node.body[i].body[0], m.ImportFrom(module=m.Name(value='__future__'))) or \
-            m.matches(updated_node.body[i].body[0], m.Expr(value=m.SimpleString()))):
+            (m.matches(updated_node.body[i].body[0], m.ImportFrom(module=m.Name(value='__future__')))):
+                # makes sure that future imports are outside the try / catch block
```
[the remark this answers is not in the document]
Comment (FlxB2) [reply src/dynapyt/instrument/CodeInstrumenter.py:107]: Should be fixed with [3bc5d56](https://github.com/sola-st/DynaPyt/pull/9/commits/3bc5d567e72fa5993389cba4c5154a12f6b0f516)

File: src/dynapyt/analyses/TraceAll.py [hunk 3]
```diff
@@ -773,7 +773,7 @@ def enter_if(self, dyn_ast: str, iid: int, cond_value: bool) -> Optional[bool]:
         """
         self.log(iid, '   If', cond_value)
 
-    def enter_for(self, dyn_ast: str, iid: int, next_value: Any, is_async: bool) -> Optional[Any]:
+    def enter_for(self, dyn_ast: str, iid: int, next_value: Any, iterable: Iterable) -> Optional[Any]:
```
Comment (FlxB2) [inline src/dynapyt/analyses/TraceAll.py:776]: Leave a comment here to warn users not to consume the iterable

File: src/dynapyt/runtime.py [hunk 4]
```diff
@@ -476,11 +476,11 @@ def _gen_(dyn_ast, iid, iterator):
     while True:
```
Comment (FlxB2) [inline src/dynapyt/runtime.py:476]: Add condition to check if iterable is None

File: src/dynapyt/run_analysis.py [hunk 5]
```diff
@@ -3,20 +3,32 @@
 from subprocess import run
 from os.path import abspath
 import sys
-from .instrument.IIDs import IIDs
+import traceback
```
[the remark this answers is not in the document]
Comment (FlxB2) [reply src/dynapyt/run_analysis.py]: Already removed :)

File: src/dynapyt/instrument/CodeInstrumenter.py [hunk 6]
```diff
@@ -92,17 +93,18 @@ def visit_AsName(self, node):
 
     # Top level
 
-    def leave_Module(self, original_node, updated_node):
+    def leave_Module(self, original_node: cst.Module, updated_node: cst.Module):
         imports_index = -1
-        parse_to_ast = cst.BinaryOperation(left=cst.Name(value='__file__'), operator=cst.Add(), right=cst.SimpleString('".orig"'))
+        abs_path = os.path.abspath(self.file_path)
+        parse_to_ast = cst.BinaryOperation(left=cst.SimpleString(value=self.__as_string(str(abs_path))), operator=cst.Add(), right=cst.SimpleString('".orig"'))
```
[the remark this answers is not in the document]
Comment (FlxB2) [reply src/dynapyt/instrument/CodeInstrumenter.py:99]: Added in Readme

File: requirements.txt [hunk 7]
```diff
@@ -1,3 +1,20 @@
...
+mypy-extensions==0.4.3
+packaging==21.3
+pluggy==1.0.0
+py==1.11.0
+pyparsing==3.0.9
+pytest==7.0.1
+pytest-forked==1.4.0
+pytest-xdist==2.5.0
+PyYAML==6.0
+tomli==1.2.3
```
[the remark this answers is not in the document]
Comment (FlxB2) [reply requirements.txt]: Done :)

**Comment by FlxB2** [comment]
@AryazE I added the changes as discussed. Tests seem to work fine as well. Let me know if anything else should be changed.

**Review by AryazE — APPROVED** [review]
LGTM

**Comment by AryazE** [comment]
Fixes #25

**Pull request merged**